pkg/operator/latencyprofilecontroller/controller.go

```diff
@@ -0,0 +1,180 @@
+	reasonLatencyProfileEmpty           = "ProfileEmpty"
+	reasonLatencyProfileSynced          = "ProfileSynced"
+	reasonLatencyProfileRollingOut      = "ProfileRollingOut"
+
+	// prefix used with status types
+	workerLatencyProfileProgressing = "WorkerLatencyProfileProgressing"
+	workerLatencyProfileComplete    = "WorkerLatencyProfileComplete"
+)
+
+type MatchProfileRevisionConfigsFunc func(profile configv1.WorkerLatencyProfileType, revisions []int32) (match bool, err error)
+
+// LatencyProfileController either instantly via the informers
+// or periodically via resync, lists the config/v1/node object
+// and fetches the worker latency profile applied on the cluster which is used to
+// updates the status of respective operator resource that uses this controller.
+// The current state of the operand is by watching the configs applied
+// to different static pod revisions that are active and uses the information to
+// update status as either progressing or completed or degraded.
+// Note: In case new latency profiles are added in the future in openshift/api
+// this could break cluster upgrades and set this controller into degraded state
+// because of an "unknown latency profile" error.
+
+type LatencyProfileController struct {
+	operatorClient   v1helpers.StaticPodOperatorClient
+	targetNamespace  string
+	configMapLister  listersv1.ConfigMapNamespaceLister
+	configNodeLister listerv1.NodeLister
+	latencyConfigs   []nodeobserver.LatencyConfigProfileTuple
+	matchRevisionsFn MatchProfileRevisionConfigsFunc
+}
+
+func NewLatencyProfileController(
+	operatorClient v1helpers.StaticPodOperatorClient,
+	targetNamespace string,
+	latencyConfigs []nodeobserver.LatencyConfigProfileTuple,
+	matchRevisionsFn MatchProfileRevisionConfigsFunc,
+	nodeInformer configv1informers.NodeInformer,
+	kubeInformersForNamespaces v1helpers.KubeInformersForNamespaces,
+	eventRecorder events.Recorder,
+) factory.Controller {
+
+	ret := &LatencyProfileController{
+		operatorClient:   operatorClient,
+		targetNamespace:  targetNamespace,
+		latencyConfigs:   latencyConfigs,
+		matchRevisionsFn: matchRevisionsFn,
+		configMapLister:  kubeInformersForNamespaces.ConfigMapLister().ConfigMaps(targetNamespace),
+		configNodeLister: nodeInformer.Lister(),
+	}
+
+	return factory.New().WithInformers(
+		// this is for our general configuration input and our status output in case another actor changes it
+		operatorClient.Informer(),
+
+		// We use nodeInformer for observing current worker latency profile
+		nodeInformer.Informer(),
+
+		// for configmaps of operator client target namespace
+		kubeInformersForNamespaces.InformersFor(targetNamespace).Core().V1().ConfigMaps().Informer(),
+	).ResyncEvery(time.Minute).WithSync(ret.sync).WithSyncDegradedOnError(operatorClient).ToController(
+		"WorkerLatencyProfile",
+		eventRecorder.WithComponentSuffix("latency-profile-controller"),
+	)
+}
+
+func (c *LatencyProfileController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
```

Comment:
are we missing some unit tests for this controller?

Reply:
follow-up is fine

Reply:
Thank you.